### ISSUE TITLE
[Bug] Entity class changing after copying outputs - and how to replicate it

### ISSUE THREAD

**Issue by Moltard** [ficool2/HammerPlusPlus-Issue-Tracker#309]
It's something that may have been reported before (outputs swapping or entity breaking), but I found the way to replicate the bug 

I reported this on the Source Engine discord yesterday

https://cdn.discordapp.com/attachments/111951182947258368/963197612519981116/bandicam_2022-04-12_00-01-39-813_new.mp4

How to break your entities in 4 steps:
1) select the trigger, go in outputs

![image](https://user-images.githubusercontent.com/65123883/162905763-81e023d7-8124-4571-8398-7d42b2386325.png)

2) copy the output targeting the entity (here vanya_ammo)

![image](https://user-images.githubusercontent.com/65123883/162905888-ee9898ad-ecb2-4fed-a7fc-b520e449868a.png)

3) double click on the output to go to that entity. You are now in the input tab

Then in the inputs tab, double click on the input to go back to the trigger entity

![image](https://user-images.githubusercontent.com/65123883/162905927-34a0abb3-5018-4895-b153-d06da3f69e12.png)

(it might be needed to click on Class info, and outputs, before going back to inputs and back to the trigger)

4) you are now on your trigger, if you click away from the property window, the trigger suddently transformed into a game player equip, which was the type of the other entity

![image](https://user-images.githubusercontent.com/65123883/162907035-2016a785-4438-402f-8bbc-88cbab050f43.png)

**note: it seems to transform it if you click away (like on a 2d view), im not sure if that can happen when closing the property window with the X button** (maybe it happened too)

the bug was able to transform a trigger_multiple into a game_player_equip, but it also was able to do the opposite. a game player equip transformed into a trigger_multiple, without any solid brush

![image](https://user-images.githubusercontent.com/65123883/162905973-da141926-0ceb-4274-a35b-9eb543d4075c.png)

And on this image, you can see the comment, which was originally belonging to the trigger multiple, so the output copying seems to carry the class + comments of entities (and maybe other stuff)

checking the vmf in notepad++ show the point entity transformed into a trigger has no solid

![image](https://user-images.githubusercontent.com/65123883/162906019-f4e56636-7fa1-4b8e-8446-660293bc97df.png)

here is a final video, showing that it easily transformed a logic_relay into a game_text, because I copied the output targeting the game text

https://cdn.discordapp.com/attachments/111951182947258368/963202832964648980/bandicam_2022-04-12_00-23-00-900_new.mp4

I hope you can fix this bug quickly, cause it's a bit problematic

**Comment by Moltard** [comment]
it may not even require to use the copy output button

going through the Inputs tab alone can possibly change entities type, by setting the type of the previous entity

**Comment by Species3259** [comment]
@Moltard Thats true.

Simple steps to reproduce entity class changing all by itself:
Hammer Build 8860 (working on Portal 2 map)

1) create 3 entities (say: prop_button / logic _case/ light)
2) click prop_button, click Outputs tab
3) click logic_case , click Class Info tab, click Outputs tab
4) click light
5) click logic_case and now your light has become a logic_case as well

Few things to notice :
- after step 4, when you click the Class Info tab (light) it states : logic_case. The rest of the tab (Property Name/Value) is blank , with a light grey background.
- it all starts when either selecting **the Outputs tab or Inputs tab first** and then switch to other entities. Just switching with the Class Info tab active functions as expected.
- all connections (Inputs/Outputs) are still there .. some broken, but unchanged otherwise.

**_So clearly, at some point when switching entities, the Class Info tab is not updated properly._**

This also works with the Mark button.
I went from a func_brush to a trigger_multiple, then on the Outputs tab i selected an Outputs-entry, clicked Mark and my trigger_multiple became a func_brush.

Ah, fun times with Hammer (++) . Out with the old and in with new (bugs :-)

But on a serious note : brilliant work (no sarcasm), love the 'new' Hammer.
Quality of Life improvement for sure !!

**Comment by WormSlayer** [comment]
The random issues caused by this bug are killing me, I'm going to have to stop using Hammer++ until its fixed :(

**Comment by Moltard** [comment]
> The random issues caused by this bug are killing me, I'm going to have to stop using Hammer++ until its fixed :(

Yeah, for me it's unusable too

The screenshot of my post come from a project that i wanted to work on with Hammer++ and it made me find the bug.

When people work with entity and outputs it's just too problematic. Guess your average competitive mapper doesnt encounter the issue 

Im back with normal Hammer for now


Though, I think you may want to try to use the build 8857, because Ficool updated the way outputs work (and probably introduced the bug) in the 8858

https://github.com/ficool2/HammerPlusPlus-Website/releases/tag/8857

**Comment by ficool2** [comment]
Fixed in Build 8861

**Comment by Moltard** [comment]
sadly not totally fixed (as shown on discord). 

I will try to find the exact method to break entity again, but from what I had tested, it was related to copying outputs

https://cdn.discordapp.com/attachments/111951182947258368/990392053085118515/bandicam_2022-06-26_01-01-38-727_new.mp4

https://cdn.discordapp.com/attachments/111951182947258368/990395123571515412/bandicam_2022-06-26_01-15-14-577_new.mp4


https://cdn.discordapp.com/attachments/111951182947258368/990395381739319346/unknown.png

**Comment by Species3259** [comment]
I cant test it myself since (at least for me) Build 8861 wont start.

**Comment by Moltard** [comment]
I found the exact method to break entities

It's very dangerous, as it involve unselecting entities while in the input tab, and selecting them again

**__How to break classname in Hammer++__**

What you need:
A **entity_2 **that has an output targeting **entity_1**
The **entity_x** can be any entity in your map. It doesnt matter

> **Step 01:** Select `entity_x`
> **Step 02:** Go in the `entity_x` input tab
> **Step 03:** Unselect the `entity_x` (click in the void in the 3D View or 2D View)
> **Step 04:** Select the `entity_1` (you are in the "Input" tab)
> **Step 05:** Double click on the source input, to go to `entity_2`
> **Step 06:** You are now in `entity_2` "Output" tab
> **Step 07:** Go to the "Class Info" tab
> **Step 08:** Go back to the "Output" tab
> **Step 09:** __/!\\__ Here you** can double click literally any of the output** available on `entity_2` , to go to the entity targeted
> **Step 10:** You are now on the **entity **that was** targetted by the output**
> **Step 11:** Go to the "Class Info" tab
> **Step 12:** __/!\\__ The **entity you are on is starting to break** and is becoming of the type of `entity_2`

Note: **the initial entity used in the step 1 to 3 could be any entity in your map**, what matters is that you unselected an entity while the input tab was open

**Comment by Species3259** [comment]
I would not call it exact since there are many ways to change an entity involuntarily.
I mentioned two of them in the above post.
I am sure even more ways can be found.

"Note: the initial entity used in the step 1 to 3 could be any entity in your map, what matters is that you unselected an entity while the input tab was open"

or the Output tab ... either way.

**Comment by Moltard** [comment]
> 

indeed, unless you are in class info tab when you unselect the entity, it is gonna break the next one

**Comment by Moltard** [comment]
I created a new issue to keep things organized (as this one is closed)

https://github.com/ficool2/HammerPlusPlus-Issue-Tracker/issues/360
